### ISSUE TITLE
This model do not support No GT in one batch case

### ISSUE THREAD

**Issue by vicdxxx** [ultralytics/yolov5#1285]
Before submitting a bug report, please be aware that your issue **must be reproducible** with all of the following, otherwise it is non-actionable, and we can not help you:
 - **Current repo**: run `git fetch && git status -uno` to check and `git pull` to update repo
 - **Common dataset**: coco.yaml or coco128.yaml
 - **Common environment**: Colab, Google Cloud, or Docker image. See https://github.com/ultralytics/yolov5#environments
 
If this is a custom dataset/training question you **must include** your `train*.jpg`, `test*.jpg` and `results.png` figures, or we can not help you. You can generate these with `utils.plot_results()`.

## 🐛 Bug

I do a test, I train 2/5 classes(only 2 class as target) use yolov5, 100 images, use batch size is 1, when it build_target() function read one image with no target class, it occur exception

please change compute_loss() / build_target() / train() relative part

I have solved this issue, it not much code, so you guys can fix this bug simply

**Comment by glenn-jocher** [comment]
@vicdxxx thanks for raising this bug report. I believe zero-target batches are correctly handled, as we've resolved older issus with this in the past, though its possible those tests were not run at batch-size 1.

Please submit a Pull Request with your updates.

**Comment by glenn-jocher** [comment]
TODO: investigate zero-target empty batch failure mode at --batch 1

**Comment by glenn-jocher** [comment]
Updates on this. I trained on COCO128, validated on COCO128 with all but two labels deleted. With all else at default settings, training proceeds nominally:

```python
Using CPU

Namespace(adam=False, batch_size=16, bucket='', cache_images=False, cfg='', data='data/coco128.yaml', device='', epochs=300, evolve=False, global_rank=-1, hyp='data/hyp.scratch.yaml', image_weights=False, img_size=[320, 320], local_rank=-1, log_imgs=10, logdir='runs/', multi_scale=False, name='', noautoanchor=False, nosave=False, notest=False, rect=False, resume=False, single_cls=False, sync_bn=False, total_batch_size=16, weights='yolov5s.pt', workers=8, world_size=1)
Start Tensorboard with "tensorboard --logdir runs/", view at http://localhost:6006/
Install Weights & Biases for experiment logging via 'pip install wandb' (recommended)
Hyperparameters {'lr0': 0.01, 'lrf': 0.2, 'momentum': 0.937, 'weight_decay': 0.0005, 'warmup_epochs': 3.0, 'warmup_momentum': 0.8, 'warmup_bias_lr': 0.1, 'box': 0.05, 'cls': 0.5, 'cls_pw': 1.0, 'obj': 1.0, 'obj_pw': 1.0, 'iou_t': 0.2, 'anchor_t': 4.0, 'fl_gamma': 0.0, 'hsv_h': 0.015, 'hsv_s': 0.7, 'hsv_v': 0.4, 'degrees': 0.0, 'translate': 0.1, 'scale': 0.5, 'shear': 0.0, 'perspective': 0.0, 'flipud': 0.0, 'fliplr': 0.5, 'mosaic': 1.0, 'mixup': 0.0}

                 from  n    params  module                                  arguments                     
  0                -1  1      3520  models.common.Focus                     [3, 32, 3]                    
  1                -1  1     18560  models.common.Conv                      [32, 64, 3, 2]                
  2                -1  1     19904  models.common.BottleneckCSP             [64, 64, 1]                   
  3                -1  1     73984  models.common.Conv                      [64, 128, 3, 2]               
  4                -1  1    161152  models.common.BottleneckCSP             [128, 128, 3]                 
  5                -1  1    295424  models.common.Conv                      [128, 256, 3, 2]              
  6                -1  1    641792  models.common.BottleneckCSP             [256, 256, 3]                 
  7                -1  1   1180672  models.common.Conv                      [256, 512, 3, 2]              
  8                -1  1    656896  models.common.SPP                       [512, 512, [5, 9, 13]]        
  9                -1  1   1248768  models.common.BottleneckCSP             [512, 512, 1, False]          
 10                -1  1    131584  models.common.Conv                      [512, 256, 1, 1]              
 11                -1  1         0  torch.nn.modules.upsampling.Upsample    [None, 2, 'nearest']          
 12           [-1, 6]  1         0  models.common.Concat                    [1]                           
 13                -1  1    378624  models.common.BottleneckCSP             [512, 256, 1, False]          
 14                -1  1     33024  models.common.Conv                      [256, 128, 1, 1]              
 15                -1  1         0  torch.nn.modules.upsampling.Upsample    [None, 2, 'nearest']          
 16           [-1, 4]  1         0  models.common.Concat                    [1]                           
 17                -1  1     95104  models.common.BottleneckCSP             [256, 128, 1, False]          
 18                -1  1    147712  models.common.Conv                      [128, 128, 3, 2]              
 19          [-1, 14]  1         0  models.common.Concat                    [1]                           
 20                -1  1    313088  models.common.BottleneckCSP             [256, 256, 1, False]          
 21                -1  1    590336  models.common.Conv                      [256, 256, 3, 2]              
 22          [-1, 10]  1         0  models.common.Concat                    [1]                           
 23                -1  1   1248768  models.common.BottleneckCSP             [512, 512, 1, False]          
 24      [17, 20, 23]  1    229245  models.yolo.Detect                      [80, [[10, 13, 16, 30, 33, 23], [30, 61, 62, 45, 59, 119], [116, 90, 156, 198, 373, 326]], [128, 256, 512]]
Model Summary: 191 layers, 7.46816e+06 parameters, 7.46816e+06 gradients

Transferred 370/370 items from yolov5s.pt
Optimizer groups: 62 .bias, 70 conv.weight, 59 other
Scanning labels ../coco128/labels/train2017.cache (126 found, 0 missing, 2 empty, 0 duplicate, for 128 images): 128it [00:00, 12400.87it/s]
Scanning labels ../coco128_copy/labels/train2017.cache (2 found, 0 missing, 126 empty, 0 duplicate, for 128 images): 128it [00:00, 135163.88it/s]

Analyzing anchors... anchors/target = 3.99, Best Possible Recall (BPR) = 0.9591. Attempting to generate improved anchors, please wait...
WARNING: Extremely small objects found. 35 of 929 labels are < 3 pixels in width or height.
Running kmeans for 9 anchors on 927 points...
thr=0.25: 0.9752 best possible recall, 3.75 anchors past thr
n=9, img_size=320, metric_all=0.261/0.655-mean/best, past_thr=0.473-mean: 8,11,  26,19,  30,47,  75,43,  52,92,  77,162,  160,107,  174,237,  299,195
Evolving anchors with Genetic Algorithm: fitness = 0.6584: 100%|██████████| 1000/1000 [00:01<00:00, 847.92it/s]
Image sizes 320 train, 320 test
Using 8 dataloader workers
Logging results to runs/exp1
Starting training for 300 epochs...

     Epoch   gpu_mem       box       obj       cls     total   targets  img_size
  0%|          | 0/8 [00:00<?, ?it/s]thr=0.25: 0.9828 best possible recall, 3.84 anchors past thr
n=9, img_size=320, metric_all=0.266/0.660-mean/best, past_thr=0.474-mean: 9,10,  25,15,  30,41,  64,43,  50,86,  69,147,  144,108,  183,196,  314,210
New anchors saved to model. Update model *.yaml to use these anchors in the future.

     0/299        0G   0.07511    0.2103   0.03091    0.3163       171       320: 100%|██████████| 8/8 [00:44<00:00,  5.54s/it]
               Class      Images     Targets           P           R      mAP@.5  mAP@.5:.95: 100%|██████████| 8/8 [00:11<00:00,  1.47s/it]
                 all         128          32      0.0185       0.554       0.196      0.0964

     Epoch   gpu_mem       box       obj       cls     total   targets  img_size
     1/299        0G   0.06334    0.1924    0.0318    0.2875       143       320: 100%|██████████| 8/8 [00:41<00:00,  5.24s/it]
               Class      Images     Targets           P           R      mAP@.5  mAP@.5:.95: 100%|██████████| 8/8 [00:11<00:00,  1.47s/it]
                 all         128          32      0.0207       0.546       0.216        0.17

     Epoch   gpu_mem       box       obj       cls     total   targets  img_size
     2/299        0G   0.05883    0.2029   0.03151    0.2933       192       320: 100%|██████████| 8/8 [00:41<00:00,  5.14s/it]
               Class      Images     Targets           P           R      mAP@.5  mAP@.5:.95: 100%|██████████| 8/8 [00:11<00:00,  1.44s/it]
                 all         128          32      0.0206       0.546       0.211        0.17
```

If I repeat with --batch 1, training proceeds nominally as well:
```python
Using CPU

Namespace(adam=False, batch_size=1, bucket='', cache_images=False, cfg='', data='data/coco128.yaml', device='', epochs=300, evolve=False, global_rank=-1, hyp='data/hyp.scratch.yaml', image_weights=False, img_size=[320, 320], local_rank=-1, log_imgs=10, logdir='runs/', multi_scale=False, name='', noautoanchor=False, nosave=False, notest=False, rect=False, resume=False, single_cls=False, sync_bn=False, total_batch_size=1, weights='yolov5s.pt', workers=8, world_size=1)
Start Tensorboard with "tensorboard --logdir runs/", view at http://localhost:6006/
Install Weights & Biases for experiment logging via 'pip install wandb' (recommended)
Hyperparameters {'lr0': 0.01, 'lrf': 0.2, 'momentum': 0.937, 'weight_decay': 0.0005, 'warmup_epochs': 3.0, 'warmup_momentum': 0.8, 'warmup_bias_lr': 0.1, 'box': 0.05, 'cls': 0.5, 'cls_pw': 1.0, 'obj': 1.0, 'obj_pw': 1.0, 'iou_t': 0.2, 'anchor_t': 4.0, 'fl_gamma': 0.0, 'hsv_h': 0.015, 'hsv_s': 0.7, 'hsv_v': 0.4, 'degrees': 0.0, 'translate': 0.1, 'scale': 0.5, 'shear': 0.0, 'perspective': 0.0, 'flipud': 0.0, 'fliplr': 0.5, 'mosaic': 1.0, 'mixup': 0.0}

                 from  n    params  module                                  arguments                     
  0                -1  1      3520  models.common.Focus                     [3, 32, 3]                    
  1                -1  1     18560  models.common.Conv                      [32, 64, 3, 2]                
  2                -1  1     19904  models.common.BottleneckCSP             [64, 64, 1]                   
  3                -1  1     73984  models.common.Conv                      [64, 128, 3, 2]               
  4                -1  1    161152  models.common.BottleneckCSP             [128, 128, 3]                 
  5                -1  1    295424  models.common.Conv                      [128, 256, 3, 2]              
  6                -1  1    641792  models.common.BottleneckCSP             [256, 256, 3]                 
  7                -1  1   1180672  models.common.Conv                      [256, 512, 3, 2]              
  8                -1  1    656896  models.common.SPP                       [512, 512, [5, 9, 13]]        
  9                -1  1   1248768  models.common.BottleneckCSP             [512, 512, 1, False]          
 10                -1  1    131584  models.common.Conv                      [512, 256, 1, 1]              
 11                -1  1         0  torch.nn.modules.upsampling.Upsample    [None, 2, 'nearest']          
 12           [-1, 6]  1         0  models.common.Concat                    [1]                           
 13                -1  1    378624  models.common.BottleneckCSP             [512, 256, 1, False]          
 14                -1  1     33024  models.common.Conv                      [256, 128, 1, 1]              
 15                -1  1         0  torch.nn.modules.upsampling.Upsample    [None, 2, 'nearest']          
 16           [-1, 4]  1         0  models.common.Concat                    [1]                           
 17                -1  1     95104  models.common.BottleneckCSP             [256, 128, 1, False]          
 18                -1  1    147712  models.common.Conv                      [128, 128, 3, 2]              
 19          [-1, 14]  1         0  models.common.Concat                    [1]                           
 20                -1  1    313088  models.common.BottleneckCSP             [256, 256, 1, False]          
 21                -1  1    590336  models.common.Conv                      [256, 256, 3, 2]              
 22          [-1, 10]  1         0  models.common.Concat                    [1]                           
 23                -1  1   1248768  models.common.BottleneckCSP             [512, 512, 1, False]          
 24      [17, 20, 23]  1    229245  models.yolo.Detect                      [80, [[10, 13, 16, 30, 33, 23], [30, 61, 62, 45, 59, 119], [116, 90, 156, 198, 373, 326]], [128, 256, 512]]
Model Summary: 191 layers, 7.46816e+06 parameters, 7.46816e+06 gradients

Transferred 370/370 items from yolov5s.pt
Optimizer groups: 62 .bias, 70 conv.weight, 59 other
Scanning labels ../coco128/labels/train2017.cache (126 found, 0 missing, 2 empty, 0 duplicate, for 128 images): 128it [00:00, 12517.10it/s]
Scanning labels ../coco128_copy/labels/train2017.cache (2 found, 0 missing, 126 empty, 0 duplicate, for 128 images): 128it [00:00, 305387.32it/s]

Analyzing anchors... anchors/target = 3.99, Best Possible Recall (BPR) = 0.9591. Attempting to generate improved anchors, please wait...
WARNING: Extremely small objects found. 35 of 929 labels are < 3 pixels in width or height.
Running kmeans for 9 anchors on 927 points...
Evolving anchors with Genetic Algorithm:   0%|          | 0/1000 [00:00<?, ?it/s]thr=0.25: 0.9752 best possible recall, 3.75 anchors past thr
n=9, img_size=320, metric_all=0.261/0.655-mean/best, past_thr=0.473-mean: 8,11,  26,19,  30,47,  75,43,  52,92,  77,162,  160,107,  174,237,  299,195
Evolving anchors with Genetic Algorithm: fitness = 0.6584: 100%|██████████| 1000/1000 [00:00<00:00, 2322.77it/s]
Image sizes 320 train, 320 test
Using 0 dataloader workers
Logging results to runs/exp2
Starting training for 300 epochs...

     Epoch   gpu_mem       box       obj       cls     total   targets  img_size
  0%|          | 0/128 [00:00<?, ?it/s]thr=0.25: 0.9828 best possible recall, 3.84 anchors past thr
n=9, img_size=320, metric_all=0.266/0.660-mean/best, past_thr=0.474-mean: 9,10,  25,15,  30,41,  64,43,  50,86,  69,147,  144,108,  183,196,  314,210
New anchors saved to model. Update model *.yaml to use these anchors in the future.

     0/299        0G   0.07229    0.2521    0.0642    0.3886        33       320: 100%|██████████| 128/128 [00:55<00:00,  2.29it/s]
               Class      Images     Targets           P           R      mAP@.5  mAP@.5:.95: 100%|██████████| 128/128 [00:10<00:00, 11.71it/s]
                 all         128          32      0.0208       0.499       0.239      0.0999

     Epoch   gpu_mem       box       obj       cls     total   targets  img_size
     1/299        0G   0.06498     0.242   0.06373    0.3707         3       320: 100%|██████████| 128/128 [00:54<00:00,  2.34it/s]
               Class      Images     Targets           P           R      mAP@.5  mAP@.5:.95: 100%|██████████| 128/128 [00:11<00:00, 11.11it/s]
                 all         128          32      0.0239       0.571       0.272       0.197

     Epoch   gpu_mem       box       obj       cls     total   targets  img_size

```

If I run test.py on my data.yaml instead of train.py, again at --batch 1 everything also operates correctly.
```python
Namespace(augment=False, batch_size=1, conf_thres=0.001, data='data/coco128.yaml', device='', img_size=320, iou_thres=0.65, save_conf=False, save_dir='runs/test', save_json=False, save_txt=False, single_cls=False, task='val', verbose=False, weights='yolov5s.pt')
Using CPU

Fusing layers... 
Model Summary: 140 layers, 7.45958e+06 parameters, 0 gradients
Scanning labels ../coco128_copy/labels/train2017.cache (2 found, 0 missing, 126 empty, 0 duplicate, for 128 images): 128it [00:00, 135334.24it/s]
               Class      Images     Targets           P           R      mAP@.5  mAP@.5:.95: 100%|██████████| 128/128 [00:10<00:00, 11.68it/s]
                 all         128          32      0.0464       0.546       0.193       0.159
Speed: 76.9/2.6/79.6 ms inference/NMS/total per 320x320 image at batch-size 1
Results saved to runs/test
```

Conclusion is that training operates correctly for zero-target validation batches regardless of --batch-size. If you believe otherwise please supply exact code to reproduce, in our colab notebook if possible. Thanks!

Removing TODO and Bug labels.
